R/allocate_reference_times.R

```diff
@@ -0,0 +1,337 @@
+      "i" = "{n_ref_times} reference times available and {size_target} are specified.", # nolint
+      "x" = "All {n_ref_times} reference times will be used." # nolint
+    ))
+    return(n_ref_times)
+  }
+
+  cli_abort(message = c(
+    "Insufficient reference times in reporting triangle for the both delay and uncertainty estimation.", # nolint
+    "i" = "{n_ref_times} reference times available and {size_required} are needed, {size_min_delay} for delay estimation and {size_min_retro_nowcasts} for uncertainty estimation.", # nolint
+    "x" = "Probabilistic nowcasts cannot be generated." # nolint
+  ))
+  return(NULL)
+}
+
+#'
+#' @inheritParams .check_against_requirements
+#' @inheritParams .assign_ns_from_sizes
+#' @inheritParams allocate_reference_times
+#'
+#' @returns NULL invisibly
+.handle_target_insufficient <- function(size_target,
+                                        size_required,
+                                        size_min_delay,
+                                        size_min_retro_nowcasts,
+                                        scale_factor,
+                                        max_delay) {
+  cli_abort(message = c(
+    "Insufficient reference times specified by `scale_factor` for the both delay and uncertainty estimation.", # nolint
+    "i" = "{scale_factor*max_delay} reference times specified and {size_required} are needed, {size_min_delay} for delay estimation and {size_min_retro_nowcasts} for uncertainty estimation.", # nolint,
+    "x" = "Probabilistic nowcasts cannot be generated." # nolint
+  ))
+  return(NULL)
+}
+
+#' Assign number of reference times to delay and uncertainty from the sizes
+#'
+#' @param size_used Integer indicating number reference times that will be used.
+#' @param size_target Integer indicating the target number of reference times.
+#' @param size_min_delay Integer indicating the number needed for delay
+#'    estimation.
+#' @param size_min_retro_nowcasts Integer indicating the number needed for
+#'    uncertainty estimation.
+#' @param size_required Integer indicating the number need for both delay and
+#'    uncertainty
+#' @inheritParams allocate_reference_times
+#'
+#' @returns List of number of reference times to use for delay and uncertainty
+.assign_ns_from_sizes <- function(size_used,
+                                  size_target,
+                                  size_min_delay,
+                                  size_min_retro_nowcasts,
+                                  size_required,
+                                  prop_delay) {
+  if (size_used >= size_target) {
+    n_history_delay <- max(floor(size_used * prop_delay), size_min_delay)
+    n_retrospective_nowcasts <- size_used - n_history_delay
+    if (n_retrospective_nowcasts < size_min_retro_nowcasts) {
+      n_retrospective_nowcasts <- size_min_retro_nowcasts
+      n_history_delay <- size_used - n_retrospective_nowcasts
```

Comment:
This line is duplicated

Reply:
I don't think the line is duplicated. This logic is implementing the case where your available reference times (size_used) are less than the target (size_target = max_delay*scale_factor)

If this is the case we first check that the delay estimation minimum is hit with:
` n_history_delay <- max(floor(size_used * prop_delay), size_min_delay)`
And if it is, then use prop delay to allocate. 

However, this needs a further check to ensure that when doing this, there are at least `size_min_retro_nowcasts`, if there aren't:
```r
 n_retrospective_nowcasts <- size_min_retro_nowcasts
 n_history_delay <- size_used - n_retrospective_nowcasts
```

Reply:
So we find delay first, assign the rest to uncertainty, and then check that uncertainty requirements are met.